### PR TITLE
Expand

### PR DESCRIPTION
**Pull request by pape-io**
LHS will ask the specific endpoint to `expand` the included linked resources on their side. If it detects that this didn't happen, it will fetch them itself.

### CODE REVIEW

File: lib/lhs/concerns/record/request.rb [hunk 1]
```diff
@@ -55,18 +55,19 @@ def extend_raw_data(data, addition, key)
...
         else
-          handle_include(includes, data)
+          handle_include!(includes, data)
         end
       end
 
-      def handle_include(included, data, sub_includes = nil)
-        return if data.blank? || skip_loading_includes?(data, included)
+      def handle_include!(included, data, sub_includes = nil)
+        return if data.blank? || skip_loading_includes?(data, included) || expanded?(included, data, sub_includes)
```
Comment (cedricwider) [inline lib/lhs/concerns/record/request.rb]: some might consider a statement like `return if a || b || c` do be worth refactoring out into a method. What do you think?

File: lib/lhs/concerns/record/request.rb [hunk 2]
```diff
@@ -146,14 +153,37 @@ def options_for_nested_items(data, key)
 
       # Merge explicit params and take configured endpoints options as base
       def process_options(options, endpoint)
+        options = options.deep_dup
         options[:params].deep_symbolize_keys! if options[:params]
         options = (endpoint.options || {}).merge(options)
         options[:url] = compute_url!(options[:params]) unless options.key?(:url)
+        options[:params] ||= {}
```
Comment (cedricwider) [inline lib/lhs/concerns/record/request.rb]: moving this statement up could save you the `if options[:params]` on line 157

**Comment by cedricwider** [comment]
added minor comments... other than that 👍

File: lib/lhs/concerns/record/request.rb (hunk 1 above)
Comment (Averethel) [inline lib/lhs/concerns/record/request.rb]: I agree with @cedricwider - factoring out condition to method would be nice

File: lib/lhs/concerns/record/request.rb [hunk 3]
```diff
@@ -146,14 +153,37 @@ def options_for_nested_items(data, key)
...
+        options[:params][:expand] = handle_expands(including) if including.present?
         merge_explicit_params!(options[:params])
         options.delete(:params) if options[:params] && options[:params].empty?
         options
       end
 
+      # Prepare expand parameter in order to get expanding done by the endpoint
+      def handle_expands(expands, prefix = nil)
+        list = []
```
Comment (Averethel) [inline lib/lhs/concerns/record/request.rb]: how about using `tap` here:
```ruby
[].tap do |list|
  ....
end.join(',')
```

File: lib/lhs/concerns/record/request.rb [hunk 4]
```diff
@@ -146,14 +153,37 @@ def options_for_nested_items(data, key)
...
+          expands.each do |element|
+            list << handle_expands(element, prefix)
+          end
+        else
+          list << [
+            prefix,
+            expands
+          ].compact.join('.')
+        end
+        list.join(',')
```
Comment (Averethel) [inline lib/lhs/concerns/record/request.rb]: Should there not be also `compact` / empty removal?

**Comment by Averethel** [comment]
Minor comments, once CI is happy so am I.

File: lib/lhs/concerns/record/request.rb (hunk 4 above)
Comment (pape-io) [inline lib/lhs/concerns/record/request.rb]: Not necessary, as the necessary lines that add stuff to the list are compacting, but will add it for comprehensibility. We want to have a cleaned list.

File: lib/lhs/concerns/record/request.rb (hunk 3 above)
Comment (pape-io) [inline lib/lhs/concerns/record/request.rb]: #tap that 👍

File: lib/lhs/concerns/record/request.rb (hunk 2 above)
Comment (pape-io) [inline lib/lhs/concerns/record/request.rb]: done 👍

**Pull request closed without merging**